src/randomenv.cpp

```diff
@@ -198,12 +198,23 @@ void AddAllCPUID(CSHA512& hasher)
     AddCPUID(hasher, 0, 0, ax, bx, cx, dx); // Returns max leaf in ax
     uint32_t max = ax;
     for (uint32_t leaf = 1; leaf <= max; ++leaf) {
+        uint32_t maxsub = 0;
         for (uint32_t subleaf = 0;; ++subleaf) {
```

Comment:
I am not an expert on CPUID, but how about an emergency break after some number of subleaves - or was it a deliberate decision to construct the inner loop in a way that it can possibly run infinitely when certain assumptions are wrong, or might be in the future?

Reply:
I guess the question is whether we want to have any chance of catching weird systems that return non-compliant CPUID values. If we add an upper bound, we'll probably never ever see a bug report like this one.

Reply:
True, I am not sure if catching those justifies the inconvenience for affected users who might not be able to run bitcoin on their system without waiting for a patch (considering that CPUID is just one of many static entropy sources).

Reply:
I think this depends on the chance that there actually CPUs out there, suitable for running nodes, that are not compliant with this. 

I feel slightly uneasy with potentially infinite loops over hardware registers like this, OTOH I understand @sipa's point.

> I guess the question is whether we want to have any chance of catching weird systems that return non-compliant CPUID values.

It could log a message when it hits the upper bound. And if you make it a high enough value you'd still get reports, just that bitcoind startup is slow instead of doesn't start at all.
(which is probably preferable :smile:)

Reply:
It's only a uint32_t. How long does a full wrap-around take?

Reply:
In this case, I'd be tempted to begin with failing painfully and adapt if needed.

Reply:
@laanwj A full 32-bit wrap-around would mean SHA512 hashing 96 GiB of data. Too long, I'm sure.

I've added a commit to limit the number of iterations.